File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.scala
object Solution {
    def solution(numbers: Vector[Int]): Vector[Int] = {
        func(numbers.zipWithIndex, Vector.empty, Vector.fill(numbers.size)(-1))
    }
    
    def func(v: Vector[(Int, Int)], r: Vector[(Int, Int)], k: Vector[Int]): Vector[Int] = {
        if (v.isEmpty) k
        else if (r.isEmpty) func(v.tail, r.appended(v.head), k)
        else if(v.head._1 > r.last._1) func(v, r.dropRight(1), k.updated(r.last._2, v.head._1))
        else func(v.tail, r.appended(v.head), k)
  }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        Stack<int[]> st = new Stack<>();
        int[] arr = new int[numbers.length];
        Arrays.fill(arr, -1);
        for (int i = 0; i < numbers.length; i++) {
            while (!st.isEmpty() && st.peek()[0] < numbers[i]) {
                arr[st.peek()[1]] = numbers[i];
                st.pop();
            }
            st.push(new int[] {numbers[i], i});
        }
        return arr;
    }
}